unledger/internal/bip39/trie.go

```diff
@@ -0,0 +1,92 @@
+package bip39
+
+import (
+	"bytes"
+	_ "embed"
+	"fmt"
+	"strings"
+)
+
+var (
+	//go:embed english.txt
+	englishWordList []byte
+	englishWordLUT  map[string]int
+	englishTrie     *trieNode
+)
+
+// ErrAmbiguous is the error returned when a mnemonic word is ambiguous.
+var ErrAmbiguous = fmt.Errorf("bip39: ambiguous mnemonic word")
+
+type trieNode struct {
+	isLeaf   bool
+	children map[byte]*trieNode
+}
+
+func (n *trieNode) Lookup(prefix string) (string, error) {
+	prefix = strings.ToLower(prefix)
+
+	var (
+		ret string
+		p   *trieNode = n
+	)
+	for _, c := range prefix {
+		if p == nil || c < 'a' || c > 'z' {
+			// Either the prefix is too long, or one or more of the
+			// characters is out of range.
+			return "", fmt.Errorf("bip39: invalid mnemonic word")
+		}
+		ret = ret + string(c)
+		p = p.children[byte(c)]
+	}
+	if p == nil || p.isLeaf {
+		return ret, nil
+	}
+	for {
+		switch len(p.children) {
+		case 0:
+			return ret, nil
+		case 1:
```

Comment:
we can probably rely on the word list not being crazy, but this doesn't check for `isLeaf` along the way

Reply:
This assumes that the word-list is well formed, yeah.

isLeaf is only uses to handle cases where the user enters a 3 or 4 (no 1 or 2 letter words in the word list) prefix that also happens to be a leaf (in which case the leaf is chosen).  Past that, isLeaf is nonsensical because that means the user entered an ambiguous prefix.

The paranoid thing to do would be to assert isLeaf is true for the 0 child case, and false for the 1 child case, so I can probably do that, and rig a test case that exercises all non-ambiguous prefixes (though that might take a while).

Reply:
Added some invariant checking, and a test case that looks up every possible prefix.